contracts/mainnet/PolymorphRootTunnel.sol

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.7.0;
+
+import "../tunnel/FxBaseRootTunnel.sol";
+import "../base/PolymorphTunnel.sol";
+import "./PolymorphRoot.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721Holder.sol";
+
+contract PolymorphRootTunnel is
+    FxBaseRootTunnel,
+    ERC721Holder,
+    PolymorphTunnel
+{
+    constructor(
+        address _checkpointManager,
+        address _fxRoot,
+        address payable _daoAddress
+    )
+        FxBaseRootTunnel(_checkpointManager, _fxRoot)
+        PolymorphTunnel(_daoAddress)
+    {}
+
+    PolymorphRoot public polymorphContract;
+
+    modifier onlyOwner(uint256 tokenId) {
+        require(
+            polymorphContract.ownerOf(tokenId) == msg.sender,
+            "Only owner can move polymorph"
+        );
+        _;
+    }
+
+    function _processMessageFromChild(bytes memory data) internal override {
+        require(
+            address(polymorphContract) != address(0),
+            "Polymorph contract hasn't been set yet"
+        );
+        (
+            uint256 tokenId,
+            address ownerAddress,
+            uint256 gene,
+            bool isNotVirgin,
+            uint256 genomeChanges
+        ) = _decodeMessage(data);
+
+        polymorphContract.approve(ownerAddress, tokenId);
+
+        polymorphContract.safeTransferFrom(
```

Comment:
Using `safeTransferFrom` causes the receiving contract to call their `onERC721Received` callback. This means that a contract can place logic for receiving there and spend more gas on behalf paid the tunnel (or hit a limitation). I'd just use `transferFrom`. Seems safe to me @strenev WDYT?

Reply:
Yes I agree, `transferFrom `should be enough and in our case there should be no need to ensure that the receiver is actually capable of receiving the NFT.